### PR TITLE
Monitor metric

### LINKED ISSUE #336: Implement the API for the metrics of the Monitor
Implement the API for the metrics of the Monitor, some basic metrics should be included. such as "vm_cpu_usage/vm_mem_usage"..

### PR DESCRIPTION
**Pull request by WangLiNaruto**
fixed #336

### CODE REVIEW

File: plugins/huaweicloud/src/main/java/org/eclipse/xpanse/orchestrator/plugin/huaweicloud/monitor/utils/HuaweiMonitorCache.java [hunk 1]
```diff
@@ -0,0 +1,76 @@
...
+
+/**
+ * Huawei Monitor Cache.
+ */
+@Component
+public class HuaweiMonitorCache {
+
+    private static final int DEFAULT_TIME_DIFFERENCE = 30 * 60 * 1000;
+
+    private static final Map<String, HuaweiMetric> metricMap = new ConcurrentHashMap<>();
```
Comment (baixinsui) [inline plugins/huaweicloud/src/main/java/org/eclipse/xpanse/orchestrator/plugin/huaweicloud/monitor/utils/HuaweiMonitorCache.java:28]: static final names 'metricMap' should use all uppercase.

File: plugins/huaweicloud/src/main/java/org/eclipse/xpanse/orchestrator/plugin/huaweicloud/monitor/utils/HuaweiMonitorClientUtil.java [hunk 2]
```diff
@@ -0,0 +1,57 @@
...
+     * @param icredential ICredential.
+     */
+    public CesClient getCesClient(ICredential icredential, String regionName) {
+        CesClient client;
+        try {
+            client = CesClient.newBuilder()
+                    .withCredential(icredential)
+                    .withRegion(CesRegion.valueOf(regionName))
+                    .build();
+        } catch (Exception e) {
```
Comment (baixinsui) [inline plugins/huaweicloud/src/main/java/org/eclipse/xpanse/orchestrator/plugin/huaweicloud/monitor/utils/HuaweiMonitorClientUtil.java:51]: It is not recommended to directly catch the basic exception class Exception

File: plugins/huaweicloud/src/main/java/org/eclipse/xpanse/orchestrator/plugin/huaweicloud/monitor/utils/HuaweiMonitorHttpConvert.java [hunk 3]
```diff
@@ -0,0 +1,98 @@
...
+            requestList.add(request);
+        }
+        return requestList;
+    }
+
+    private static String getNameSpace() {
+        return HuaweiNameSpaceKind.ECS_SYS.toValue();
+    }
+
+    private static String getMetricName(String resourceKind) {
```
Comment (baixinsui) [inline plugins/huaweicloud/src/main/java/org/eclipse/xpanse/orchestrator/plugin/huaweicloud/monitor/utils/HuaweiMonitorHttpConvert.java:62]: The param of this method can use  the enum MonitorResourceType  directly.

File: plugins/huaweicloud/src/main/java/org/eclipse/xpanse/orchestrator/plugin/huaweicloud/monitor/utils/HuaweiMonitorHttpConvert.java [hunk 4]
```diff
@@ -0,0 +1,98 @@
...
+        metric.setLabels(labels);
+        metric.setType(MetricType.GAUGE);
+
+        List<Datapoint> datapoints = response.getDatapoints();
+        if (!CollectionUtils.isEmpty(datapoints)) {
+            datapoints.sort(Comparator.comparing(Datapoint::getTimestamp).reversed());
+            MetricItem metricItem = new MetricItem();
+            metricItem.setType(MetricItemType.VALUE);
+            metricItem.setValue(datapoints.get(0).getAverage());
+            metric.setMetrics(Arrays.asList(metricItem));
```
Comment (iskey) [inline plugins/huaweicloud/src/main/java/org/eclipse/xpanse/orchestrator/plugin/huaweicloud/monitor/utils/HuaweiMonitorHttpConvert.java:93]:
```suggestion
            metric.setMetrics(List.of(metricItem));
```

File: plugins/huaweicloud/src/main/java/org/eclipse/xpanse/orchestrator/plugin/huaweicloud/monitor/utils/HuaweiMonitorHttpConvert.java [hunk 5]
```diff
@@ -0,0 +1,98 @@
...
+
+        Metric metric = new Metric();
+        metric.setName(request.getMetricName());
+        Map<String, String> labels = new HashMap<>();
+        labels.put("id", deployResource.getResourceId());
+        labels.put("name", deployResource.getName());
+        metric.setLabels(labels);
+        metric.setType(MetricType.GAUGE);
+
+        List<Datapoint> datapoints = response.getDatapoints();
```
Comment (iskey) [inline plugins/huaweicloud/src/main/java/org/eclipse/xpanse/orchestrator/plugin/huaweicloud/monitor/utils/HuaweiMonitorHttpConvert.java:87]:
```suggestion
        List<Datapoint> datapointList = response.getDatapoints();
```

File: plugins/huaweicloud/src/main/java/org/eclipse/xpanse/orchestrator/plugin/huaweicloud/monitor/utils/HuaweiMonitorHttpConvert.java [hunk 6]
```diff
@@ -0,0 +1,98 @@
...
+/**
+ * Resource conversion.
+ */
+@Slf4j
+public class HuaweiMonitorHttpConvert {
+
+    /**
+     * Get HuaweiCloud Monitor ShowMetricDataRequest.
+     */
+    public static List<ShowMetricDataRequest> convertToMetricDataRequest(
```
Comment (iskey) [inline plugins/huaweicloud/src/main/java/org/eclipse/xpanse/orchestrator/plugin/huaweicloud/monitor/utils/HuaweiMonitorHttpConvert.java:41]:
```suggestion
    public static List<ShowMetricDataRequest> buildMetricDataRequest(
```

File: plugins/huaweicloud/src/main/java/org/eclipse/xpanse/orchestrator/plugin/huaweicloud/monitor/utils/HuaweiMonitorHttpConvert.java [hunk 7]
```diff
@@ -0,0 +1,98 @@
...
+import org.springframework.util.CollectionUtils;
+
+/**
+ * Resource conversion.
+ */
+@Slf4j
+public class HuaweiMonitorHttpConvert {
+
+    /**
+     * Get HuaweiCloud Monitor ShowMetricDataRequest.
```
Comment (iskey) [inline plugins/huaweicloud/src/main/java/org/eclipse/xpanse/orchestrator/plugin/huaweicloud/monitor/utils/HuaweiMonitorHttpConvert.java:39]:
```suggestion
     * Build ShowMetricDataRequest for HuaweiCloud Monitor client.
```

File: modules/api/src/main/java/org/eclipse/xpanse/api/MonitorApi.java [hunk 8]
```diff
@@ -0,0 +1,54 @@
...
+
+    /**
+     * Get Monitor Metric.
+     *
+     * @param id Service ID.
+     */
+    @Tag(name = "Monitor",
+            description = "APIs to Monitor Metric.")
+    @Operation(description = "Get Monitor metric of the deployed service ")
+    @GetMapping(value = "/monitor/metric/{id}", produces = MediaType.APPLICATION_JSON_VALUE)
```
Comment (iskey) [inline modules/api/src/main/java/org/eclipse/xpanse/api/MonitorApi.java:47]: should we make the metrics types query parameters？
such as `cpu_util`,`mem_util`?

**Review by iskey — CHANGES_REQUESTED** [review]
add some comments.

**Review by iskey — CHANGES_REQUESTED** [review]
some comments added.

File: plugins/huaweicloud/src/main/java/org/eclipse/xpanse/orchestrator/plugin/huaweicloud/monitor/constant/HuaweiMonitorConstant.java [hunk 9]
```diff
@@ -0,0 +1,24 @@
...
+ * SPDX-FileCopyrightText: Huawei Inc.
+ *
+ */
+
+package org.eclipse.xpanse.orchestrator.plugin.huaweicloud.monitor.constant;
+
+/**
+ * Huawei Monitor Constant.
+ */
+public class HuaweiMonitorConstant {
```
Comment (swaroopar) [inline plugins/huaweicloud/src/main/java/org/eclipse/xpanse/orchestrator/plugin/huaweicloud/monitor/constant/HuaweiMonitorConstant.java:12]: class name must be plural.. HuaweiMonitorConstants

File: plugins/huaweicloud/src/main/java/org/eclipse/xpanse/orchestrator/plugin/huaweicloud/monitor/models/HuaweiMonitorMetric.java [hunk 10]
```diff
@@ -0,0 +1,21 @@
...
+ * SPDX-FileCopyrightText: Huawei Inc.
+ *
+ */
+
+package org.eclipse.xpanse.orchestrator.plugin.huaweicloud.monitor.models;
+
+/**
+ * Huawei Monitor Metric.
+ */
+public class HuaweiMonitorMetric {
```
Comment (swaroopar) [inline plugins/huaweicloud/src/main/java/org/eclipse/xpanse/orchestrator/plugin/huaweicloud/monitor/models/HuaweiMonitorMetric.java:12]: HuaweiMonitorMetric must also be plural

File: plugins/huaweicloud/src/main/java/org/eclipse/xpanse/orchestrator/plugin/huaweicloud/monitor/models/HuaweiMetric.java [hunk 11]
```diff
@@ -0,0 +1,31 @@
...
+import java.util.List;
+import lombok.Getter;
+import lombok.Setter;
+import org.eclipse.xpanse.modules.monitor.Metric;
+
+/**
+ * Huawei Metric.
+ */
+
+public class HuaweiMetric {
```
Comment (swaroopar) [inline plugins/huaweicloud/src/main/java/org/eclipse/xpanse/orchestrator/plugin/huaweicloud/monitor/models/HuaweiMetric.java:18]: Please rename all classes to have word HuaweiCloud instead of just Huawei

File: plugins/huaweicloud/src/main/java/org/eclipse/xpanse/orchestrator/plugin/huaweicloud/monitor/models/HuaweiMonitorMetric.java (hunk 10 above)
Comment (swaroopar) [inline plugins/huaweicloud/src/main/java/org/eclipse/xpanse/orchestrator/plugin/huaweicloud/monitor/models/HuaweiMonitorMetric.java:12]: remove unused constants.

File: plugins/huaweicloud/src/main/java/org/eclipse/xpanse/orchestrator/plugin/huaweicloud/monitor/models/HuaweiMonitorMetric.java [hunk 12]
```diff
@@ -0,0 +1,21 @@
...
+
+    //cpu
+    public static final String CPU_USAGE = "cpu_usage";
+    public static final String CPU_UTIL = "cpu_util";
```
Comment (swaroopar) [inline plugins/huaweicloud/src/main/java/org/eclipse/xpanse/orchestrator/plugin/huaweicloud/monitor/models/HuaweiMonitorMetric.java:16]: should be full name.. CPU_UTILIZED and MEM_UTILIZED.

If possible these metrics can be a enum also.

File: plugins/huaweicloud/src/main/java/org/eclipse/xpanse/orchestrator/plugin/huaweicloud/monitor/utils/HuaweiMonitorHttpConvert.java [hunk 13]
```diff
@@ -0,0 +1,98 @@
...
+import org.eclipse.xpanse.orchestrator.plugin.huaweicloud.monitor.constant.HuaweiMonitorConstant;
+import org.eclipse.xpanse.orchestrator.plugin.huaweicloud.monitor.models.HuaweiMonitorMetric;
+import org.eclipse.xpanse.orchestrator.plugin.huaweicloud.monitor.models.HuaweiNameSpaceKind;
+import org.springframework.util.CollectionUtils;
+
+/**
+ * Resource conversion.
+ */
+@Slf4j
+public class HuaweiMonitorHttpConvert {
```
Comment (swaroopar) [inline plugins/huaweicloud/src/main/java/org/eclipse/xpanse/orchestrator/plugin/huaweicloud/monitor/utils/HuaweiMonitorHttpConvert.java:36]: class name can be HuaweiCloudToXpanseDataModelConverter

File: plugins/huaweicloud/src/main/java/org/eclipse/xpanse/orchestrator/plugin/huaweicloud/monitor/utils/HuaweiMonitorClientUtil.java [hunk 14]
```diff
@@ -0,0 +1,58 @@
...
+    public CesClient getCesClient(ICredential icredential, String regionName) {
+        CesClient client;
+        try {
+            client = CesClient.newBuilder()
+                    .withCredential(icredential)
+                    .withRegion(CesRegion.valueOf(regionName))
+                    .build();
+        } catch (SdkException e) {
+            log.error("Get HuaweiCloud Monitor CesClient failed.");
+            return null;
```
Comment (swaroopar) [inline plugins/huaweicloud/src/main/java/org/eclipse/xpanse/orchestrator/plugin/huaweicloud/monitor/utils/HuaweiMonitorClientUtil.java:54]: should we throw an exception here? otherwise the calling method will have null pointers

File: plugins/huaweicloud/src/main/java/org/eclipse/xpanse/orchestrator/plugin/huaweicloud/monitor/utils/HuaweiMonitorCache.java [hunk 15]
```diff
@@ -0,0 +1,76 @@
...
+            return new ArrayList<>();
+        }
+        return huaweiMetric.getMetrics();
+    }
+
+    /**
+     * Cache Expires (Default 30 minutes).
+     */
+    @Async("taskExecutor")
+    public void expire() {
```
Comment (swaroopar) [inline plugins/huaweicloud/src/main/java/org/eclipse/xpanse/orchestrator/plugin/huaweicloud/monitor/utils/HuaweiMonitorCache.java:66]: we must add MDC keys to this method. And add a log line that the old metrices are being cleared.

File: plugins/huaweicloud/src/main/java/org/eclipse/xpanse/orchestrator/plugin/huaweicloud/monitor/utils/HuaweiMonitorClientUtil.java [hunk 16]
```diff
@@ -0,0 +1,58 @@
...
+    /**
+     * Get Huawei Monitor ICredential by IAM.
+     *
+     * @param ak huawei ak.
+     * @param sk huawei sk.
+     */
+    public ICredential getIcredential(String ak, String sk) {
+        if (StringUtils.isBlank(ak) || StringUtils.isBlank(sk)) {
+            log.error("Get ICredential error,AK or SK is empty");
+            return null;
```
Comment (swaroopar) [inline plugins/huaweicloud/src/main/java/org/eclipse/xpanse/orchestrator/plugin/huaweicloud/monitor/utils/HuaweiMonitorClientUtil.java:34]: should we throw an exception here?

File: plugins/huaweicloud/src/main/java/org/eclipse/xpanse/orchestrator/plugin/huaweicloud/monitor/utils/HuaweiMonitorClientUtil.java [hunk 17]
```diff
@@ -0,0 +1,58 @@
...
+public class HuaweiMonitorClientUtil {
+
+    /**
+     * Get Huawei Monitor ICredential by IAM.
+     *
+     * @param ak huawei ak.
+     * @param sk huawei sk.
+     */
+    public ICredential getIcredential(String ak, String sk) {
+        if (StringUtils.isBlank(ak) || StringUtils.isBlank(sk)) {
```
Comment (swaroopar) [inline plugins/huaweicloud/src/main/java/org/eclipse/xpanse/orchestrator/plugin/huaweicloud/monitor/utils/HuaweiMonitorClientUtil.java:32]: we are setting ak and sk to null initially. Should we also test for null?

File: plugins/huaweicloud/src/main/java/org/eclipse/xpanse/orchestrator/plugin/huaweicloud/monitor/utils/HuaweiMonitorClientUtil.java [hunk 18]
```diff
@@ -0,0 +1,58 @@
...
+@Component
+public class HuaweiMonitorClientUtil {
+
+    /**
+     * Get Huawei Monitor ICredential by IAM.
+     *
+     * @param ak huawei ak.
+     * @param sk huawei sk.
+     */
+    public ICredential getIcredential(String ak, String sk) {
```
Comment (swaroopar) [inline plugins/huaweicloud/src/main/java/org/eclipse/xpanse/orchestrator/plugin/huaweicloud/monitor/utils/HuaweiMonitorClientUtil.java:31]: would prefer to use full names instead of shortforms ak and sk

File: modules/api/src/main/java/org/eclipse/xpanse/api/MonitorApi.java (hunk 8 above)
Comment (swaroopar) [reply modules/api/src/main/java/org/eclipse/xpanse/api/MonitorApi.java:47]: yes good idea. and better to have this an enum.

File: modules/api/src/main/java/org/eclipse/xpanse/api/MonitorApi.java [hunk 19]
```diff
@@ -0,0 +1,54 @@
...
+    @Resource
+    private OrchestratorService orchestratorService;
+
+    /**
+     * Get Monitor Metric.
+     *
+     * @param id Service ID.
+     */
+    @Tag(name = "Monitor",
+            description = "APIs to Monitor Metric.")
```
Comment (swaroopar) [inline modules/api/src/main/java/org/eclipse/xpanse/api/MonitorApi.java:45]: APIs to Monitor Metric ->please use either of the two as description.

1. APIs to get metrics of deployed services. 
2. APIs to monitor deployed services.

File: modules/api/src/main/java/org/eclipse/xpanse/api/MonitorApi.java [hunk 20]
```diff
@@ -0,0 +1,54 @@
...
+    @Operation(description = "Get Monitor metric of the deployed service ")
```
Comment (swaroopar) [inline modules/api/src/main/java/org/eclipse/xpanse/api/MonitorApi.java:46]: description - Get metrics of the deployed service.

File: modules/orchestrator/src/main/java/org/eclipse/xpanse/orchestrator/OrchestratorService.java [hunk 21]
```diff
@@ -330,4 +337,17 @@ public Deployment getDeployment(DeployerKind deployerKind) {
         return deployment;
     }
 
+    /**
+     * Get Monitor Metric.
+     *
+     * @param id service id.
+     */
+    public List<Metric> getMetrics(UUID id) {
```
Comment (swaroopar) [inline modules/orchestrator/src/main/java/org/eclipse/xpanse/orchestrator/OrchestratorService.java:345]: getMetricsByServiceId

**Review by swaroopar — COMMENTED** [review]
comments added. please also consider adding unit tests for wherever possible. (either with same issue or a new issue)

**Review by baixinsui — APPROVED** [review]
Looks good to me.

File: plugins/huaweicloud/src/main/java/org/eclipse/xpanse/orchestrator/plugin/huaweicloud/HuaweiCloudOrchestratorPlugin.java [hunk 22]
```diff
@@ -49,23 +67,84 @@ public List<CredentialType> getAvailableCredentialTypes() {
...
     }
 
     @Override
     public List<Metric> getMetrics(AbstractCredentialInfo credential,
-            DeployResource deployResource) {
-        return null;
+            DeployResource deployResource, MonitorResourceType monitorResourceType) {
+        List<Metric> metrics = new ArrayList<>();
+        try {
+            clearExpiredMetricCache(deployResource.getResourceId());
```
Comment (swaroopar) [inline plugins/huaweicloud/src/main/java/org/eclipse/xpanse/orchestrator/plugin/huaweicloud/HuaweiCloudOrchestratorPlugin.java:90]: since we return data for the past "x" minutes only.. we must have this documented in the interface. If this, could be different for each cloud, we must document it accordingly in the API.

**Review by swaroopar — CHANGES_REQUESTED** [review]
looks mostly good. just two small comments.

File: plugins/huaweicloud/src/main/java/org/eclipse/xpanse/orchestrator/plugin/huaweicloud/HuaweiCloudOrchestratorPlugin.java (hunk 22 above)
Comment (WangLiNaruto) [reply plugins/huaweicloud/src/main/java/org/eclipse/xpanse/orchestrator/plugin/huaweicloud/HuaweiCloudOrchestratorPlugin.java:89]: We do not fetch data from the past period of time, we obtain real-time data. However, when a scheduled task calls HUAWEI CLOUD monitoring data, if the call is frequent, the interface will limit the flow, so there will be a buffer solution. When the HUAWEI CLOUD monitoring interface reports an error , we will fetch the data in the cache, and the monitoring data of each vm in the cache will be invalidated in 30 minutes, but these invalidated data will be cleared every 12 hours

**Review by swaroopar — APPROVED** [review]

**Review by iskey — APPROVED** [review]
looks good.
